pkg/listwatch/listwatch.go

```diff
@@ -0,0 +1,179 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Provides multiple namespace listerWatcher. This implementation is Largely from
+// https://github.com/coreos/prometheus-operator/pkg/listwatch/listwatch.go
+
+package listwatch
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/tools/cache"
+)
+
+// MultiNamespaceListerWatcher takes a list of namespaces and a
+// cache.ListerWatcher generator func and returns a single cache.ListerWatcher
+// capable of operating on multiple namespaces.
+func MultiNamespaceListerWatcher(namespaces []string, f func(string) cache.ListerWatcher) cache.ListerWatcher {
+	// If there is only one namespace then there is no need to create a proxy.
+	if len(namespaces) == 1 {
+		return f(namespaces[0])
+	}
+	var lws []cache.ListerWatcher
+	for _, n := range namespaces {
+		lws = append(lws, f(n))
+	}
+	return multiListerWatcher(lws)
+}
+
+// multiListerWatcher abstracts several cache.ListerWatchers, allowing them
+// to be treated as a single cache.ListerWatcher.
+type multiListerWatcher []cache.ListerWatcher
+
+// List implements the ListerWatcher interface.
+// It combines the output of the List method of every ListerWatcher into
+// a single result.
+func (mlw multiListerWatcher) List(options metav1.ListOptions) (runtime.Object, error) {
+	l := metav1.List{}
+	var resourceVersions []string
+	for _, lw := range mlw {
+		list, err := lw.List(options)
+		if err != nil {
+			return nil, err
+		}
+		items, err := meta.ExtractList(list)
+		if err != nil {
+			return nil, err
+		}
+		metaObj, err := meta.ListAccessor(list)
+		if err != nil {
+			return nil, err
+		}
+		for _, item := range items {
+			l.Items = append(l.Items, runtime.RawExtension{Object: item.DeepCopyObject()})
+		}
+		resourceVersions = append(resourceVersions, metaObj.GetResourceVersion())
+	}
+	// Combine the resource versions so that the composite Watch method can
+	// distribute appropriate versions to each underlying Watch func.
+	l.ListMeta.ResourceVersion = strings.Join(resourceVersions, "/")
+	return &l, nil
+}
+
+// Watch implements the ListerWatcher interface.
+// It returns a watch.Interface that combines the output from the
+// watch.Interface of every cache.ListerWatcher into a single result chan.
+func (mlw multiListerWatcher) Watch(options metav1.ListOptions) (watch.Interface, error) {
+	resourceVersions := make([]string, len(mlw))
+	// Allow resource versions to be "".
+	if options.ResourceVersion != "" {
```

Comment:
This is extremely counter intuitive and breaks the regular contract of options. Not supporting a starting resource version (i.e. erroring out) is a simpler choice here.